test/e2e/ocpbugs48050_test.go

```diff
@@ -0,0 +1,611 @@
+	}
+
+	prometheusClient, err := metrics.NewPrometheusClient(context.TODO(), kubeClient, routeClient)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create Prometheus client: %v", err)
+	}
+
+	return prometheusClient, nil
+}
+
+func queryPrometheus(client prometheusv1client.API, query string, timeout time.Duration) (model.Value, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), timeout)
+	defer cancel()
+
+	result, warnings, err := client.Query(ctx, query, time.Now())
+	if err != nil {
+		return nil, fmt.Errorf("error querying Prometheus: %v", err)
+	}
+
+	if len(warnings) > 0 {
+		return nil, fmt.Errorf("warnings querying Prometheus: %v", warnings)
+	}
+
+	return result, nil
+}
+
+type podScrapeCounts map[string]float64
+
+func getTotalScrapesForPods(promClient prometheusv1client.API, timeout time.Duration) (podScrapeCounts, error) {
+	query := `sum by (pod) (haproxy_exporter_total_scrapes{pod=~"router-default-.*"})`
+
+	result, err := queryPrometheus(promClient, query, timeout)
+	if err != nil {
+		return nil, err
+		//t.Fatalf("Failed to query Prometheus for total scrapes: %v", err)
+	}
+
+	vector, ok := result.(model.Vector)
+	if !ok {
+		return nil, fmt.Errorf("expected result type model.Vector, got %T", result)
+	}
+
+	scrapeCounts := make(map[string]float64)
+
+	for _, sample := range vector {
+		podName := string(sample.Metric["pod"])
+		scrapeCounts[podName] = float64(sample.Value)
+	}
+
+	return scrapeCounts, nil
+}
+
+func waitForDefaultRouterScrapesIncrement(promClient prometheusv1client.API, minNewScrapes int, timeout time.Duration, progress func(initialCounts, currentCounts podScrapeCounts, allIncremented bool)) error {
+	initialCounts, err := getTotalScrapesForPods(promClient, timeout)
+	if err != nil {
+		return fmt.Errorf("failed to get initial scrape counts: %v", err)
+	}
+
+	deadline := time.Now().Add(timeout)
+
+	for time.Now().Before(deadline) {
+		currentCounts, err := getTotalScrapesForPods(promClient, timeout)
+		if err != nil {
+			return fmt.Errorf("failed to get current scrape counts: %v", err)
+		}
+
+		allIncremented := true
+		for podName, initialCount := range initialCounts {
+			if currentCount, exists := currentCounts[podName]; !exists || currentCount < initialCount+float64(minNewScrapes) {
+				allIncremented = false
+				break
+			}
+		}
+
+		// Call the progress function with the current state
+		if progress != nil {
+			progress(initialCounts, currentCounts, allIncremented)
+		}
+
+		if allIncremented {
+			return nil
+		}
+
+		time.Sleep(5 * time.Second)
+	}
+
+	return fmt.Errorf("timeout after %v waiting for all pods to increment total scrapes by at least %d", timeout, minNewScrapes)
+}
+
+func getCanaryImageFromIngressOperatorDeployment() (string, error) {
+	ingressOperator := types.NamespacedName{Namespace: operatorNamespace, Name: "ingress-operator"}
+
+	deployment := appsv1.Deployment{}
+	if err := kclient.Get(context.TODO(), ingressOperator, &deployment); err != nil {
```

Comment:
I see many cases where we don't use polling; for example, here's one from `canary_test.go`:

```go
	// Get default ingress controller deployment
	deployment := &appsv1.Deployment{}
	if err := kclient.Get(context.TODO(), controller.RouterDeploymentName(def), deployment); err != nil {
		t.Fatalf("failed to get ingresscontroller deployment: %v", err)
	}
```

If there were simultaneous updates to the resource, I'd consider adding retry logic. However, since we create this resource specifically for the test and it's only used within the test, is retrying really necessary? What other code would be updating this resource in this context?

Reply:
We do:

```go
func TestOCPBUGS48050(t *testing.T) {
	if err := waitForIngressControllerCondition(t, kclient, 5*time.Minute, defaultName, defaultAvailableConditions...); err != nil {
		t.Fatalf("failed to observe expected conditions: %v", err)
	}
```

before starting the the test. Is that sufficient given that `TestOCPBUGS48050` is run serially?

Reply:
I can move this to using `waitForDeploymentComplete`.

Reply:
> Is that sufficient given that TestOCPBUGS48050 is run serially?

Probably yes. Since we don't do updates after, we don't risk much (of getting update errors).